pkg/apis/run/v1alpha1/runstatus_types.go

```diff
@@ -0,0 +1,144 @@
+type RunStatus struct {
+	duckv1.Status `json:",inline"`
+
+	// RunStatusFields inlines the status fields.
+	RunStatusFields `json:",inline"`
+}
+
+// +k8s:deepcopy-gen=true
+
+// RunStatusFields holds the fields of Run's status.  This is defined
+// separately and inlined so that other types can readily consume these fields
+// via duck typing.
+type RunStatusFields struct {
+	// StartTime is the time the build is actually started.
+	// +optional
+	StartTime *metav1.Time `json:"startTime,omitempty"`
+
+	// CompletionTime is the time the build completed.
+	// +optional
+	CompletionTime *metav1.Time `json:"completionTime,omitempty"`
+
+	// Results reports any output result values to be consumed by later
+	// tasks in a pipeline.
+	// +optional
+	Results []RunResult `json:"results,omitempty"`
+
+	// ExtraFields holds arbitrary fields provided by the custom task
+	// controller.
+	ExtraFields runtime.RawExtension `json:"extraFields,omitempty"`
+}
+
+// RunResult used to describe the results of a task
+type RunResult struct {
+	// Name the given name
+	Name string `json:"name"`
+	// Value the given value of the result
+	Value string `json:"value"`
+}
+
+var runCondSet = apis.NewBatchConditionSet()
+
+// GetCondition returns the Condition matching the given type.
+func (r *RunStatus) GetCondition(t apis.ConditionType) *apis.Condition {
+	return runCondSet.Manage(r).GetCondition(t)
+}
+
+// InitializeConditions will set all conditions in runCondSet to unknown for the PipelineRun
+// and set the started time to the current time
+func (r *RunStatus) InitializeConditions() {
+	started := false
+	if r.StartTime.IsZero() {
+		r.StartTime = &metav1.Time{Time: time.Now()}
+		started = true
+	}
+	conditionManager := runCondSet.Manage(r)
+	conditionManager.InitializeConditions()
+	// Ensure the started reason is set for the "Succeeded" condition
+	if started {
+		initialCondition := conditionManager.GetCondition(apis.ConditionSucceeded)
+		initialCondition.Reason = "Started"
+		conditionManager.SetCondition(*initialCondition)
+	}
+}
+
+// SetCondition sets the condition, unsetting previous conditions with the same
+// type as necessary.
+func (r *RunStatus) SetCondition(newCond *apis.Condition) {
+	if newCond != nil {
+		runCondSet.Manage(r).SetCondition(*newCond)
+	}
+}
+
+// MarkRunSucceeded changes the Succeeded condition to True with the provided reason and message.
+func (r *RunStatus) MarkRunSucceeded(reason, messageFormat string, messageA ...interface{}) {
+	runCondSet.Manage(r).MarkTrueWithReason(apis.ConditionSucceeded, reason, messageFormat, messageA...)
+	succeeded := r.GetCondition(apis.ConditionSucceeded)
+	r.CompletionTime = &succeeded.LastTransitionTime.Inner
+}
+
+// MarkRunFailed changes the Succeeded condition to False with the provided reason and message.
+func (r *RunStatus) MarkRunFailed(reason, messageFormat string, messageA ...interface{}) {
+	runCondSet.Manage(r).MarkFalse(apis.ConditionSucceeded, reason, messageFormat, messageA...)
```

Comment:
Should this be MarkFalseWithReason to match MarkTrueWithReason above? 🤔

Reply:
Yeah I had noticed that but this is how the functions are defined in the knative package.

```
	// MarkTrue sets the status of t to true, and then marks the happy condition to
	// true if all dependents are true.
	MarkTrue(t ConditionType)

	// MarkTrueWithReason sets the status of t to true with the reason, and then marks the happy
	// condition to true if all dependents are true.
	MarkTrueWithReason(t ConditionType, reason, messageFormat string, messageA ...interface{})

	// MarkUnknown sets the status of t to Unknown and also sets the happy condition
	// to Unknown if no other dependent condition is in an error state.
	MarkUnknown(t ConditionType, reason, messageFormat string, messageA ...interface{})

	// MarkFalse sets the status of t and the happy condition to False.
	MarkFalse(t ConditionType, reason, messageFormat string, messageA ...interface{})
```